cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll

```diff
@@ -142,28 +142,15 @@ private class LogicalNotValueNumber extends ValueNumber {
 /**
  * A Boolean condition in the AST that guards one or more basic blocks. This includes
  * operands of logical operators but not switch statements.
- *
- * For performance reasons conditions inside static local initializers or
- * global initializers are not considered `GuardCondition`s.
  */
-cached
-class GuardCondition extends Expr {
-  cached
-  GuardCondition() {
-    exists(IRGuardCondition ir | this = ir.getUnconvertedResultExpression())
-    or
-    // no binary operators in the IR
-    this.(BinaryLogicalOperation).getAnOperand() instanceof GuardCondition
-  }
```

Comment:
Why was this removed?

Reply:
Before the introduction of `final` aliases this class couldn't be made `abstract` (since it would expose a abstract class from our libraries to queries). So instead of having this class be abstract we followed the pattern of:
1. Specifying each subclass of this "pseudo-abstract" class in the charpred
2. Provide a subclass that extended this "pseudo-abstract" class and repeated the required charpred to define the subclass.

That is, we've had to do:
```qll
class Base {
  Base() {
     myImpl1(this) or myImpl2(this)
  }
}

class MyImpl1 extends Base {
  MyImpl1() {
    myImpl1(this)
  }
}

class MyImpl2 extends Base {
  MyImpl2() {
    myImpl1(this)
  }
}
```

Now that we have final classes we can get rid of this pattern (as we've done many other places) and actually use abstract classes. And abstract classes don't need this charpred since they're defined as the union of all the subclasses.

That is, we can rewrite the above example to:
```qll
private abstract class BaseImpl { }

final class Base = BaseImpl;

class MyImpl1 extends BaseImpl {
  MyImpl1() {
    myImpl1(this)
  }
}

class MyImpl2 extends BaseImpl {
  MyImpl2() {
    myImpl1(this)
  }
}
```

Does that make sense?

Reply:
Makes total sense.